File: src/main/java/org/vanautrui/languages/compiler/parsing/astnodes/typenodes/TypeParameterNode.java
package org.vanautrui.languages.compiler.parsing.astnodes.typenodes;

import org.vanautrui.languages.compiler.parsing.astnodes.typenodes.basic_and_wrapped.IBasicAndWrappedTypeNode;

public final class TypeParameterNode implements IBasicAndWrappedTypeNode, ITypeNode {

	public int typeParameterIndex;

	public TypeParameterNode(){}

	public TypeParameterNode(final int typeParameterIndex){
		this.typeParameterIndex=typeParameterIndex;
	}

	@Override
	public String toSourceCode() {
		return "?T" + this.typeParameterIndex;
	}

	@Override
	public String getTypeName() {
		return this.toSourceCode();
	}

	@Override
	public boolean isPrimitive() {
		//TODO: fix this. we do not know. we return false, but the real type could be primitive.
		return false;
	}
}

File: dragon-parser/src/main/compiler/parsing/astnodes/typenodes/TypeParameterNode.cpp

//standard headers
#include <vector>
#include <set>

//project headers
#include "TypeParameterNode.hpp"
//	  compiler/parsing/astnodes/typenodes/TypeNode.cpp
#include "../../../lexing/TokenList.hpp"

class TypeParameterNode : IBasicAndWrappedTypeNode, ITypeNode {

public:
	int typeParameterIndex;

	TypeParameterNode(TokenList tokens) throws Exception {
		IToken token = tokens.get(0);
		if (token instanceof TypeParameterIdentifierToken) {
			this.typeParameterIndex = ((TypeParameterIdentifierToken) token).type_parameter_number;
			tokens.consume(1);
		} else {
			throw new Exception("could not read type parameter node");
		}
	}

};